### PR TITLE
Further pyclient types and refactoring

### PR DESCRIPTION
**Pull request by dtebbs**
Preparation for exposing zeth functionality as CLI tools.
- Big refactor of tests into calls to a single "joinsplit()' operation.
- Stricter types.
- Some naming more consistent with Zet paper

The cli tools are not complete, but this is a good place to merge, to avoid branches diverging too much.

### CODE REVIEW

File: pyClient/zeth/utils.py [hunk 1]
```diff
@@ -14,13 +14,21 @@
...
-# Wei).
-ZETH_PUBLIC_UNIT_VALUE = 1000000000000
-
 W3 = Web3(HTTPProvider(constants.WEB3_HTTP_PROVIDER))
 
 
+class EtherValue:
+    """
+    Representation of some amount of ether (or any token) in terms of Wei.
+    Disambiguates ether values from other units such as zeth_units.
```
Comment (AntoineRondelet) [inline pyClient/zeth/utils.py]: Ether

File: pyClient/zeth/zksnark.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+# pylint: disable=unnecessary-pass
+
+# Dictionary representing a VerificationKey from any supported snark
+GenericVerificationKey = Dict[str, Any]
+
+# Dictionary representing a Proof from any supported snark
+GenericProof = Dict[str, Any]
+
+
+class IZKSnarkProvider(ABC):
```
Comment (AntoineRondelet) [inline pyClient/zeth/zksnark.py]: Please add a line of comments to explain that this class represents the Interface that needs to be implemented by all snark providers

File: pyClient/zeth/joinsplit.py [hunk 3]
```diff
@@ -634,7 +650,27 @@ def trap_r_randomness() -> str:
     return bytes(Random.get_random_bytes(48)).hex()
 
 
-def _signature_randomness() -> bytes:
+def _compute_rho_i(phi: str, hsig: bytes, i: int) -> bytes:
+    """
+    Returns rho_i = blake2s(0 || i || 10 || [phi]_252 || hsig)
+    See: Zcash protocol spec p. 57, Section 5.4.2 Pseudo Random Functions
+    """
+    # [SANITY CHECK] make sure i is in the interval [0, 1]
```
Comment (AntoineRondelet) [inline pyClient/zeth/joinsplit.py]: `interval [0, 1]` -> `interval [0, JS_INPUTS]`

File: pyClient/zeth/joinsplit.py [hunk 4]
```diff
@@ -292,64 +270,34 @@ def encode_pub_input_to_hash(message_list: List[Union[str, List[str]]]) -> bytes
...
-
-    return res
-
-
-def sign(
-        keypair: JoinsplitKeypair,
-        hash_ciphers: str,
-        hash_proof: str,
-        hash_inputs: str) -> int:
+def _sign(
```
Comment (AntoineRondelet) [inline pyClient/zeth/joinsplit.py]: This needs to be moved to `signing.py` which should contain all the algorithms of the signature scheme we use. Namely, `Sign(sk, msg)` (returns `sigma`), `Verify(vk, sigma, msg)` (return 0/1), `KeyGen(\secparam)` (returns `sk, vk`).

Thus, the signature of `_sign` above function needs to be modified (cc: @rrtoledo as this is related to the thread:https://github.com/clearmatics/zeth/pull/122#discussion_r356465339)

File: pyClient/zeth/joinsplit.py [hunk 5]
```diff
@@ -1,88 +1,124 @@
...
+    """
+    Secret addr_sk, consisting of a_sk and k_sk
+    """
+    def __init__(self, a_sk: OwnershipSecretKey, k_sk: EncryptionSecretKey):
+        self.a_sk: OwnershipSecretKey = a_sk
+        self.k_sk: EncryptionSecretKey = k_sk
 
 
-class JoinsplitKeypair:
+class JoinsplitKeyPair:
```
Comment (AntoineRondelet) [inline pyClient/zeth/joinsplit.py]: Rename to `ZethAddress`.

`JoinsplitKeyPair`  does not capture the essence of this data structure (which does not characterize a given joinsplit, but trully is a Zeth address) and conflicts with the pair of signing keys is also referred to as `JoinsplitSigKey` (Section 4.6 of the Zcash specs) and is such that:
```
joinSplitPrivKey ← JoinSplitSig.GenPrivate()
joinSplitPubKey := JoinSplitSig.DerivePublic( joinSplitPrivKey )
```

File: pyClient/zeth/joinsplit.py [hunk 6]
```diff
@@ -1,88 +1,124 @@
...
+    Public half of a zethAddress.  addr_pk = (a_pk and k_pk)
+    """
+    def __init__(self, a_pk: OwnershipPublicKey, k_pk: EncryptionPublicKey):
+        self.a_pk: OwnershipPublicKey = a_pk
+        self.k_pk: EncryptionPublicKey = k_pk
 
 
-# Joinsplit Public Key
-JoinsplitPublicKey = Tuple[G1, G1]
+class JoinsplitSecretKey:
```
Comment (AntoineRondelet) [inline pyClient/zeth/joinsplit.py]: Rename these classes, to `AddrPub`/`AddrPriv` (i.e. public and private address of a ZethAddress, see comment below)

File: pyClient/test_commands/test_ether_mixing.py [hunk 7]
```diff
@@ -238,21 +192,15 @@ def main() -> None:
...
         # recompiute the path to have the updated nodes
-        mk_byte_tree = get_merkle_tree(mixer_instance)
-        mk_path = zeth.utils.compute_merkle_path(
-            cm_address_bob_to_charlie2, mk_tree_depth, mk_byte_tree)
         result_double_spending = scenario.charlie_double_withdraw(
-            prover_client,
-            mixer_instance,
+            zeth_client,
+            # prover_client,
+            # mixer_instance,
```
Comment (AntoineRondelet) [inline pyClient/test_commands/test_ether_mixing.py]: Shall we remove these comments?

File: pyClient/zeth/ownership.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+OwnershipPublicKey = NewType('OwnershipPublicKey', bytes)
+
+
+class OwnershipKeyPair:
+    """
+    Key-pair for ownership proof
+    """
+    def __init__(self, a_sk: OwnershipSecretKey, a_pk: OwnershipPublicKey):
+        self.a_pk: OwnershipPublicKey = a_pk
+        self.a_sk: OwnershipSecretKey = a_sk
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply pyClient/zeth/ownership.py]: Moved this

File: pyClient/zeth/utils.py (hunk 1 above)
Comment (dtebbs) [reply pyClient/zeth/utils.py]: Changed

File: pyClient/zeth/zksnark.py (hunk 2 above)
Comment (dtebbs) [reply pyClient/zeth/zksnark.py]: Add a comment for this.

File: pyClient/zeth/joinsplit.py (hunk 3 above)
Comment (dtebbs) [reply pyClient/zeth/joinsplit.py]: Updated

File: pyClient/zeth/joinsplit.py (hunk 5 above)
Comment (dtebbs) [reply pyClient/zeth/joinsplit.py]: Renamed (see comment below)

File: pyClient/zeth/joinsplit.py (hunk 6 above)
Comment (dtebbs) [reply pyClient/zeth/joinsplit.py]: Renamed to `ZethAddress`, `ZethAddressPub` and `ZethAddressPriv` to keep consistency.

File: pyClient/test_commands/test_ether_mixing.py (hunk 7 above)
Comment (dtebbs) [reply pyClient/test_commands/test_ether_mixing.py]: Removed.

**Pull request merged**